#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-40_iptv.dpatch by Tobias Grimm <tg@e-tobi.net>
## from version 0.0.3 of the IPTV plug-in
## http://www.saunalahti.fi/~rahrenbe/vdr/iptv/
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Patch required for the IPTV plugin-in

@DPATCH@
diff -urNad vdr-1.4.7~/channels.c vdr-1.4.7/channels.c
--- vdr-1.4.7~/channels.c	2006-05-28 17:03:40.000000000 +0200
+++ vdr-1.4.7/channels.c	2007-10-29 19:37:57.000000000 +0100
@@ -166,6 +166,7 @@
   shortName = strdup("");
   provider = strdup("");
   portalName = strdup("");
+  pluginParam = strdup("");
   memset(&__BeginData__, 0, (char *)&__EndData__ - (char *)&__BeginData__);
   inversion    = INVERSION_AUTO;
   bandwidth    = BANDWIDTH_AUTO;
@@ -187,6 +188,7 @@
   shortName = NULL;
   provider = NULL;
   portalName = NULL;
+  pluginParam = NULL;
   schedule     = NULL;
   linkChannels = NULL;
   refChannel   = NULL;
@@ -215,6 +217,7 @@
   free(shortName);
   free(provider);
   free(portalName);
+  free(pluginParam);
 }
 
 cChannel& cChannel::operator= (const cChannel &Channel)
@@ -223,6 +226,7 @@
   shortName = strcpyrealloc(shortName, Channel.shortName);
   provider = strcpyrealloc(provider, Channel.provider);
   portalName = strcpyrealloc(portalName, Channel.portalName);
+  pluginParam = strcpyrealloc(pluginParam, Channel.pluginParam);
   memcpy(&__BeginData__, &Channel.__BeginData__, (char *)&Channel.__EndData__ - (char *)&Channel.__BeginData__);
   return *this;
 }
@@ -280,9 +284,26 @@
      transmission = Channel->transmission;
      guard        = Channel->guard;
      hierarchy    = Channel->hierarchy;
+     if (IsPlug())  pluginParam = strcpyrealloc(pluginParam, Channel->pluginParam);
      }
 }
 
+bool cChannel::SetPlugTransponderData(int Source, int Frequency, const char *PluginParam)
+{
+  if (source != Source || frequency != Frequency || (strcmp(pluginParam, PluginParam) != 0)) {
+       if (Number()) {
+          dsyslog("changing transponder data of channel %d from %s:%d:%s to %s:%d:%s", Number(), *cSource::ToString(source), frequency, pluginParam, *cSource::ToString(Source), Frequency, PluginParam);
+          modification |= CHANNELMOD_TRANSP;
+          Channels.SetModified();
+          }
+       source = Source;
+       frequency = Frequency;
+       pluginParam = strcpyrealloc(pluginParam, PluginParam);
+       schedule = NULL;
+       }
+    return true;
+}
+
 bool cChannel::SetSatTransponderData(int Source, int Frequency, char Polarization, int Srate, int CoderateH)
 {
   // Workarounds for broadcaster stupidity:
@@ -407,6 +428,18 @@
      }
 }
 
+void cChannel::SetPluginParam(const char *PluginParam)
+{
+  if (!isempty(PluginParam) && strcmp(pluginParam, PluginParam) != 0) {
+     if (Number()) {
+        dsyslog("changing plugin parameters of channel %d from '%s' to '%s'", Number(), pluginParam, PluginParam);
+        modification |= CHANNELMOD_TRANSP;
+        Channels.SetModified();
+        }
+     pluginParam = strcpyrealloc(pluginParam, PluginParam);
+     }
+}
+
 #define STRDIFF 0x01
 #define VALDIFF 0x02
 
@@ -579,7 +612,7 @@
   if (isdigit(type))
      type = 'S';
 #define ST(s) if (strchr(s, type))
-  char buffer[64];
+  char buffer[256];
   char *q = buffer;
   *q = 0;
   ST(" S ")  q += sprintf(q, "%c", polarization);
@@ -591,6 +624,7 @@
   ST("  T")  q += PrintParameter(q, 'T', MapToUser(transmission, TransmissionValues));
   ST("  T")  q += PrintParameter(q, 'G', MapToUser(guard, GuardValues));
   ST("  T")  q += PrintParameter(q, 'Y', MapToUser(hierarchy, HierarchyValues));
+  ST("P  ")  snprintf(buffer, sizeof(buffer), "%s", pluginParam);
   return buffer;
 }
 
@@ -612,7 +646,7 @@
 
 bool cChannel::StringToParameters(const char *s)
 {
-  while (s && *s) {
+  while (s && *s && !IsPlug()) {
         switch (toupper(*s)) {
           case 'B': s = ParseParameter(s, bandwidth, BandwidthValues); break;
           case 'C': s = ParseParameter(s, coderateH, CoderateValues); break;
@@ -722,7 +756,7 @@
         dpids[0] = 0;
         ok = false;
         if (parambuf && sourcebuf && vpidbuf && apidbuf) {
-           ok = StringToParameters(parambuf) && (source = cSource::FromString(sourcebuf)) >= 0;
+           ok = ((source = cSource::FromString(sourcebuf)) >= 0) && StringToParameters(parambuf);
 
            char *p = strchr(vpidbuf, '+');
            if (p)
@@ -813,6 +847,7 @@
            shortName = strcpyrealloc(shortName, p);
            }
         name = strcpyrealloc(name, namebuf);
+        if (IsPlug()) pluginParam = strcpyrealloc(pluginParam, parambuf);
 
         free(parambuf);
         free(sourcebuf);
diff -urNad vdr-1.4.7~/channels.h vdr-1.4.7/channels.h
--- vdr-1.4.7~/channels.h	2006-05-28 17:03:56.000000000 +0200
+++ vdr-1.4.7/channels.h	2007-10-29 19:37:57.000000000 +0100
@@ -114,6 +114,7 @@
   char *shortName;
   char *provider;
   char *portalName;
+  char *pluginParam;
   int __BeginData__;
   int frequency; // MHz
   int source;
@@ -165,6 +166,7 @@
   int Frequency(void) const { return frequency; } ///< Returns the actual frequency, as given in 'channels.conf'
   int Transponder(void) const;                    ///< Returns the transponder frequency in MHz, plus the polarization in case of sat
   static int Transponder(int Frequency, char Polarization); ///< builds the transponder from the given Frequency and Polarization
+  const char *PluginParam(void) const { return pluginParam; }
   int Source(void) const { return source; }
   int Srate(void) const { return srate; }
   int Vpid(void) const { return vpid; }
@@ -198,6 +200,7 @@
   int Hierarchy(void) const { return hierarchy; }
   const cLinkChannels* LinkChannels(void) const { return linkChannels; }
   const cChannel *RefChannel(void) const { return refChannel; }
+  bool IsPlug(void) const { return cSource::IsPlug(source); }
   bool IsCable(void) const { return cSource::IsCable(source); }
   bool IsSat(void) const { return cSource::IsSat(source); }
   bool IsTerr(void) const { return cSource::IsTerr(source); }
@@ -205,12 +208,14 @@
   bool HasTimer(void) const;
   int Modification(int Mask = CHANNELMOD_ALL);
   void CopyTransponderData(const cChannel *Channel);
+  bool SetPlugTransponderData(int Source, int Frequency, const char *PluginParam);
   bool SetSatTransponderData(int Source, int Frequency, char Polarization, int Srate, int CoderateH);
   bool SetCableTransponderData(int Source, int Frequency, int Modulation, int Srate, int CoderateH);
   bool SetTerrTransponderData(int Source, int Frequency, int Bandwidth, int Modulation, int Hierarchy, int CodeRateH, int CodeRateL, int Guard, int Transmission);
   void SetId(int Nid, int Tid, int Sid, int Rid = 0);
   void SetName(const char *Name, const char *ShortName, const char *Provider);
   void SetPortalName(const char *PortalName);
+  void SetPluginParam(const char *PluginParam);
   void SetPids(int Vpid, int Ppid, int *Apids, char ALangs[][MAXLANGCODE2], int *Dpids, char DLangs[][MAXLANGCODE2], int Tpid);
   void SetCaIds(const int *CaIds); // list must be zero-terminated
   void SetCaDescriptors(int Level);
diff -urNad vdr-1.4.7~/config.h vdr-1.4.7/config.h
--- vdr-1.4.7~/config.h	2007-05-12 11:07:16.000000000 +0200
+++ vdr-1.4.7/config.h	2007-10-29 19:38:07.000000000 +0100
@@ -29,6 +29,10 @@
 #define APIVERSION  "1.4.5"
 #define APIVERSNUM   10405  // Version * 10000 + Major * 100 + Minor
 
+#define PLUGINPARAMPATCHVERSNUM 1
+
+#define CLOSEFILTERPATCHVERSNUM 1
+
 // When loading plugins, VDR searches them by their APIVERSION, which
 // may be smaller than VDRVERSION in case there have been no changes to
 // VDR header files since the last APIVERSION. This allows compiled
diff -urNad vdr-1.4.7~/config.h.orig vdr-1.4.7/config.h.orig
--- vdr-1.4.7~/config.h.orig	1970-01-01 01:00:00.000000000 +0100
+++ vdr-1.4.7/config.h.orig	2007-10-29 19:37:57.000000000 +0100
@@ -0,0 +1,266 @@
+/*
+ * config.h: Configuration file handling
+ *
+ * See the main source file 'vdr.c' for copyright information and
+ * how to reach the author.
+ *
+ * $Id: config.h 1.282.1.5 2007/05/12 09:07:16 kls Exp $
+ */
+
+#ifndef __CONFIG_H
+#define __CONFIG_H
+
+#include <arpa/inet.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+#include "i18n.h"
+#include "tools.h"
+
+// VDR's own version number:
+
+#define VDRVERSION  "1.4.7"
+#define VDRVERSNUM   10407  // Version * 10000 + Major * 100 + Minor
+
+// The plugin API's version number:
+
+#define APIVERSION  "1.4.5"
+#define APIVERSNUM   10405  // Version * 10000 + Major * 100 + Minor
+
+#define PLUGINPARAMPATCHVERSNUM 1
+
+// When loading plugins, VDR searches them by their APIVERSION, which
+// may be smaller than VDRVERSION in case there have been no changes to
+// VDR header files since the last APIVERSION. This allows compiled
+// plugins to work with newer versions of the core VDR as long as no
+// VDR header files have changed.
+
+#define MAXPRIORITY 99
+#define MAXLIFETIME 99
+
+#define MINOSDWIDTH  480
+#define MAXOSDWIDTH  672
+#define MINOSDHEIGHT 324
+#define MAXOSDHEIGHT 567
+
+#define MaxFileName 256
+#define MaxSkinName 16
+#define MaxThemeName 16
+
+class cCommand : public cListObject {
+private:
+  char *title;
+  char *command;
+  bool confirm;
+  static char *result;
+public:
+  cCommand(void);
+  virtual ~cCommand();
+  bool Parse(const char *s);
+  const char *Title(void) { return title; }
+  bool Confirm(void) { return confirm; }
+  const char *Execute(const char *Parameters = NULL);
+  };
+
+typedef uint32_t in_addr_t; //XXX from /usr/include/netinet/in.h (apparently this is not defined on systems with glibc < 2.2)
+
+class cSVDRPhost : public cListObject {
+private:
+  struct in_addr addr;
+  in_addr_t mask;
+public:
+  cSVDRPhost(void);
+  bool Parse(const char *s);
+  bool Accepts(in_addr_t Address);
+  };
+
+template<class T> class cConfig : public cList<T> {
+private:
+  char *fileName;
+  bool allowComments;
+  void Clear(void)
+  {
+    free(fileName);
+    fileName = NULL;
+    cList<T>::Clear();
+  }
+public:
+  cConfig(void) { fileName = NULL; }
+  virtual ~cConfig() { free(fileName); }
+  const char *FileName(void) { return fileName; }
+  bool Load(const char *FileName = NULL, bool AllowComments = false, bool MustExist = false)
+  {
+    cConfig<T>::Clear();
+    if (FileName) {
+       free(fileName);
+       fileName = strdup(FileName);
+       allowComments = AllowComments;
+       }
+    bool result = !MustExist;
+    if (fileName && access(fileName, F_OK) == 0) {
+       isyslog("loading %s", fileName);
+       FILE *f = fopen(fileName, "r");
+       if (f) {
+          char *s;
+          int line = 0;
+          cReadLine ReadLine;
+          result = true;
+          while ((s = ReadLine.Read(f)) != NULL) {
+                line++;
+                if (allowComments) {
+                   char *p = strchr(s, '#');
+                   if (p)
+                      *p = 0;
+                   }
+                stripspace(s);
+                if (!isempty(s)) {
+                   T *l = new T;
+                   if (l->Parse(s))
+                      Add(l);
+                   else {
+                      esyslog("ERROR: error in %s, line %d", fileName, line);
+                      delete l;
+                      result = false;
+                      break;
+                      }
+                   }
+                }
+          fclose(f);
+          }
+       else {
+          LOG_ERROR_STR(fileName);
+          result = false;
+          }
+       }
+    if (!result)
+       fprintf(stderr, "vdr: error while reading '%s'\n", fileName);
+    return result;
+  }
+  bool Save(void)
+  {
+    bool result = true;
+    T *l = (T *)this->First();
+    cSafeFile f(fileName);
+    if (f.Open()) {
+       while (l) {
+             if (!l->Save(f)) {
+                result = false;
+                break;
+                }
+             l = (T *)l->Next();
+             }
+       if (!f.Close())
+          result = false;
+       }
+    else
+       result = false;
+    return result;
+  }
+  };
+
+class cCommands : public cConfig<cCommand> {};
+
+class cSVDRPhosts : public cConfig<cSVDRPhost> {
+public:
+  bool Acceptable(in_addr_t Address);
+  };
+
+extern cCommands Commands;
+extern cCommands RecordingCommands;
+extern cSVDRPhosts SVDRPhosts;
+
+class cSetupLine : public cListObject {
+private:
+  char *plugin;
+  char *name;
+  char *value;
+public:
+  cSetupLine(void);
+  cSetupLine(const char *Name, const char *Value, const char *Plugin = NULL);
+  virtual ~cSetupLine();
+  virtual int Compare(const cListObject &ListObject) const;
+  const char *Plugin(void) { return plugin; }
+  const char *Name(void) { return name; }
+  const char *Value(void) { return value; }
+  bool Parse(char *s);
+  bool Save(FILE *f);
+  };
+
+class cSetup : public cConfig<cSetupLine> {
+  friend class cPlugin; // needs to be able to call Store()
+private:
+  void StoreLanguages(const char *Name, int *Values);
+  bool ParseLanguages(const char *Value, int *Values);
+  bool Parse(const char *Name, const char *Value);
+  cSetupLine *Get(const char *Name, const char *Plugin = NULL);
+  void Store(const char *Name, const char *Value, const char *Plugin = NULL, bool AllowMultiple = false);
+  void Store(const char *Name, int Value, const char *Plugin = NULL);
+public:
+  // Also adjust cMenuSetup (menu.c) when adding parameters here!
+  int __BeginData__;
+  int OSDLanguage;
+  char OSDSkin[MaxSkinName];
+  char OSDTheme[MaxThemeName];
+  int PrimaryDVB;
+  int ShowInfoOnChSwitch;
+  int TimeoutRequChInfo;
+  int MenuScrollPage;
+  int MenuScrollWrap;
+  int MenuButtonCloses;
+  int MarkInstantRecord;
+  char NameInstantRecord[MaxFileName];
+  int InstantRecordTime;
+  int LnbSLOF;
+  int LnbFrequLo;
+  int LnbFrequHi;
+  int DiSEqC;
+  int SetSystemTime;
+  int TimeSource;
+  int TimeTransponder;
+  int MarginStart, MarginStop;
+  int AudioLanguages[I18nNumLanguages + 1];
+  int EPGLanguages[I18nNumLanguages + 1];
+  int EPGScanTimeout;
+  int EPGBugfixLevel;
+  int EPGLinger;
+  int SVDRPTimeout;
+  int ZapTimeout;
+  int PrimaryLimit;
+  int DefaultPriority, DefaultLifetime;
+  int PausePriority, PauseLifetime;
+  int UseSubtitle;
+  int UseVps;
+  int VpsMargin;
+  int RecordingDirs;
+  int VideoDisplayFormat;
+  int VideoFormat;
+  int UpdateChannels;
+  int UseDolbyDigital;
+  int ChannelInfoPos;
+  int ChannelInfoTime;
+  int OSDLeft, OSDTop, OSDWidth, OSDHeight;
+  int OSDMessageTime;
+  int UseSmallFont;
+  int MaxVideoFileSize;
+  int SplitEditedFiles;
+  int MinEventTimeout, MinUserInactivity;
+  int MultiSpeedMode;
+  int ShowReplayMode;
+  int ResumeID;
+  int CurrentChannel;
+  int CurrentVolume;
+  int CurrentDolby;
+  int InitialChannel;
+  int InitialVolume;
+  int __EndData__;
+  cSetup(void);
+  cSetup& operator= (const cSetup &s);
+  bool Load(const char *FileName);
+  bool Save(void);
+  };
+
+extern cSetup Setup;
+
+#endif //__CONFIG_H
diff -urNad vdr-1.4.7~/device.c vdr-1.4.7/device.c
--- vdr-1.4.7~/device.c	2006-09-03 12:13:25.000000000 +0200
+++ vdr-1.4.7/device.c	2007-10-29 19:38:07.000000000 +0100
@@ -509,6 +509,11 @@
   return -1;
 }
 
+void cDevice::CloseFilter(int Handle)
+{
+  close(Handle);
+}
+
 void cDevice::AttachFilter(cFilter *Filter)
 {
   if (sectionHandler)
diff -urNad vdr-1.4.7~/device.h vdr-1.4.7/device.h
--- vdr-1.4.7~/device.h	2006-06-15 11:32:48.000000000 +0200
+++ vdr-1.4.7/device.h	2007-10-29 19:38:07.000000000 +0100
@@ -302,6 +302,11 @@
        ///< Opens a file handle for the given filter data.
        ///< A derived device that provides section data must
        ///< implement this function.
+  virtual void CloseFilter(int Handle);
+       ///< Closes a file handle that has previously been opened
+       ///< by OpenFilter(). If this is as simple as calling close(Handle),
+       ///< a derived class need not implement this function, because this
+       ///< is done by the default implementation.
   void AttachFilter(cFilter *Filter);
        ///< Attaches the given filter to this device.
   void Detach(cFilter *Filter);
diff -urNad vdr-1.4.7~/dvbdevice.c vdr-1.4.7/dvbdevice.c
--- vdr-1.4.7~/dvbdevice.c	2007-02-24 12:10:14.000000000 +0100
+++ vdr-1.4.7/dvbdevice.c	2007-10-29 19:38:07.000000000 +0100
@@ -726,6 +726,11 @@
   return -1;
 }
 
+void cDvbDevice::CloseFilter(int Handle)
+{
+  close(Handle);
+}
+
 void cDvbDevice::TurnOffLiveMode(bool LiveView)
 {
   if (LiveView) {
diff -urNad vdr-1.4.7~/dvbdevice.h vdr-1.4.7/dvbdevice.h
--- vdr-1.4.7~/dvbdevice.h	2006-05-28 17:05:19.000000000 +0200
+++ vdr-1.4.7/dvbdevice.h	2007-10-29 19:38:07.000000000 +0100
@@ -79,6 +79,7 @@
 
 protected:
   virtual int OpenFilter(u_short Pid, u_char Tid, u_char Mask);
+  virtual void CloseFilter(int Handle);
 
 // Image Grab facilities
 
diff -urNad vdr-1.4.7~/i18n.c vdr-1.4.7/i18n.c
--- vdr-1.4.7~/i18n.c	2007-02-25 13:08:26.000000000 +0100
+++ vdr-1.4.7/i18n.c	2007-10-29 19:37:57.000000000 +0100
@@ -6126,6 +6126,28 @@
     "Ingen titel",
     "Bez názvu",
   },
+  { "Parameters",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "Parametrit",
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+    "",// TODO
+  },
   { NULL }
   };
 
diff -urNad vdr-1.4.7~/menu.c vdr-1.4.7/menu.c
--- vdr-1.4.7~/menu.c	2006-12-02 12:12:02.000000000 +0100
+++ vdr-1.4.7/menu.c	2007-10-29 19:37:57.000000000 +0100
@@ -216,6 +216,7 @@
   cChannel *channel;
   cChannel data;
   char name[256];
+  char pluginParam[256];
   void Setup(void);
 public:
   cMenuEditChannel(cChannel *Channel, bool New = false);
@@ -248,6 +249,7 @@
 
   // Parameters for all types of sources:
   strn0cpy(name, data.name, sizeof(name));
+  strn0cpy(pluginParam, data.pluginParam, sizeof(pluginParam));
   Add(new cMenuEditStrItem( tr("Name"),          name, sizeof(name), tr(FileNameChars)));
   Add(new cMenuEditSrcItem( tr("Source"),       &data.source));
   Add(new cMenuEditIntItem( tr("Frequency"),    &data.frequency));
@@ -276,6 +278,7 @@
   ST("  T")  Add(new cMenuEditMapItem( tr("Transmission"), &data.transmission, TransmissionValues));
   ST("  T")  Add(new cMenuEditMapItem( tr("Guard"),        &data.guard,        GuardValues));
   ST("  T")  Add(new cMenuEditMapItem( tr("Hierarchy"),    &data.hierarchy,    HierarchyValues, tr("none")));
+  ST("P  ")  Add(new cMenuEditStrItem( tr("Parameters"),    pluginParam, sizeof(pluginParam), tr(FileNameChars)));
 
   SetCurrent(Get(current));
   Display();
@@ -290,6 +293,7 @@
      if (Key == kOk) {
         if (Channels.HasUniqueChannelID(&data, channel)) {
            data.name = strcpyrealloc(data.name, name);
+           data.pluginParam = strcpyrealloc(data.pluginParam, pluginParam);
            if (channel) {
               *channel = data;
               isyslog("edited channel %d %s", channel->Number(), *data.ToText());
diff -urNad vdr-1.4.7~/sections.c vdr-1.4.7/sections.c
--- vdr-1.4.7~/sections.c	2005-09-18 13:30:29.000000000 +0200
+++ vdr-1.4.7/sections.c	2007-10-29 19:38:07.000000000 +0100
@@ -105,7 +105,7 @@
   for (fh = filterHandles.First(); fh; fh = filterHandles.Next(fh)) {
       if (fh->filterData.Is(FilterData->pid, FilterData->tid, FilterData->mask)) {
          if (--fh->used <= 0) {
-            close(fh->handle);
+            device->CloseFilter(fh->handle);
             filterHandles.Del(fh);
             break;
             }
diff -urNad vdr-1.4.7~/sources.c vdr-1.4.7/sources.c
--- vdr-1.4.7~/sources.c	2004-12-26 12:58:52.000000000 +0100
+++ vdr-1.4.7/sources.c	2007-10-29 19:37:57.000000000 +0100
@@ -37,6 +37,7 @@
   char buffer[16];
   char *q = buffer;
   switch (Code & st_Mask) {
+    case stPlug:  *q++ = 'P'; break;
     case stCable: *q++ = 'C'; break;
     case stSat:   *q++ = 'S';
                   {
@@ -56,6 +57,7 @@
 {
   int type = stNone;
   switch (toupper(*s)) {
+    case 'P': type = stPlug;  break;
     case 'C': type = stCable; break;
     case 'S': type = stSat;   break;
     case 'T': type = stTerr;  break;
diff -urNad vdr-1.4.7~/sources.conf vdr-1.4.7/sources.conf
--- vdr-1.4.7~/sources.conf	2007-02-17 17:15:13.000000000 +0100
+++ vdr-1.4.7/sources.conf	2007-10-29 19:37:57.000000000 +0100
@@ -188,3 +188,7 @@
 # Terrestrial
 
 T       Terrestrial
+
+# Plugin
+
+P       Plugin
diff -urNad vdr-1.4.7~/sources.h vdr-1.4.7/sources.h
--- vdr-1.4.7~/sources.h	2005-05-14 11:30:41.000000000 +0200
+++ vdr-1.4.7/sources.h	2007-10-29 19:37:57.000000000 +0100
@@ -16,10 +16,11 @@
 public:
   enum eSourceType {
     stNone  = 0x0000,
+    stPlug  = 0x2000,
     stCable = 0x4000,
     stSat   = 0x8000,
     stTerr  = 0xC000,
-    st_Mask = 0xC000,
+    st_Mask = 0xE000,
     st_Neg  = 0x0800,
     st_Pos  = 0x07FF,
     };
@@ -35,6 +36,7 @@
   static cString ToString(int Code);
   static int FromString(const char *s);
   static int FromData(eSourceType SourceType, int Position = 0, bool East = false);
+  static bool IsPlug(int Code) { return (Code & st_Mask) == stPlug; }
   static bool IsCable(int Code) { return (Code & st_Mask) == stCable; }
   static bool IsSat(int Code) { return (Code & st_Mask) == stSat; }
   static bool IsTerr(int Code) { return (Code & st_Mask) == stTerr; }
